runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkInputDataProcessor.java

```diff
@@ -0,0 +1,395 @@
+              continue; // try to consume outputIterator from start of loop
+            }
+            DoFnInvokers.invokerFor(ctx.getDoFn()).invokeTeardown();
+            return endOfData();
+          }
+        }
+      } catch (final RuntimeException re) {
+        DoFnInvokers.invokerFor(ctx.getDoFn()).invokeTeardown();
+        throw re;
+      }
+    }
+
+    private void fireTimer(TimerInternals.TimerData timer) {
+      StateNamespace namespace = timer.getNamespace();
+      checkArgument(namespace instanceof StateNamespaces.WindowNamespace);
+      BoundedWindow window = ((StateNamespaces.WindowNamespace) namespace).getWindow();
+      ctx.getDoFnRunner()
+          .onTimer(
+              timer.getTimerId(),
+              timer.getTimerFamilyId(),
+              ctx.getKey(),
+              window,
+              timer.getTimestamp(),
+              timer.getOutputTimestamp(),
+              timer.getDomain());
+    }
+  }
+}
+
+class AsyncSparkInputDataProcessor<FnInputT, FnOutputT>
+    implements SparkInputDataProcessor<FnInputT, FnOutputT, Tuple2<TupleTag<?>, WindowedValue<?>>> {
+
+  private final BlockingDoFnOutputManager outputManager = new BlockingDoFnOutputManager();
+
+  @Override
+  public OutputManager getOutputManager() {
+    return outputManager;
+  }
+
+  @Override
+  public <K> Iterable<Tuple2<TupleTag<?>, WindowedValue<?>>> process(
+      Iterator<WindowedValue<FnInputT>> input, SparkProcessContext<K, FnInputT, FnOutputT> ctx) {
+    if (!input.hasNext()) {
+      return Collections.emptyList();
+    }
+    return () -> new AsyncInOutIterator<>(input, ctx);
+  }
+
+  /**
+   * Output manager which can hold limited number of output elements. If capacity is reached, then
+   * attempt to output more elements will block until some elements are consumed.
+   */
+  private class BlockingDoFnOutputManager
+      implements OutputManager, Iterable<Tuple2<TupleTag<?>, WindowedValue<?>>> {
+
+    private final LinkedBlockingQueue<Tuple2<TupleTag<?>, WindowedValue<?>>> queue =
+        new LinkedBlockingQueue<>(500);
+    private volatile boolean stopped = false;
+
+    public void stop() {
+      stopped = true;
+    }
+
+    @Override
+    public Iterator<Tuple2<TupleTag<?>, WindowedValue<?>>> iterator() {
+      return new Iterator<Tuple2<TupleTag<?>, WindowedValue<?>>>() {
+
+        @Override
+        public boolean hasNext() {
+          // expect elements appearing in queue until stop() is invoked.
+          // after that, no more inputs can arrive, so just drain the queue
+          while (true) {
+            if (queue.isEmpty()) {
+              try {
+                // Wait for a bit before checking again if more data is available
+                Thread.sleep(20);
+                if (stopped) {
+                  return !queue.isEmpty();
+                }
+              } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
```

Comment:
This should break the loop. Might be safer to rethrow the exception, because terminating the iteration looks dangerous.